### PR TITLE
[SPARK-14419] [SQL] Improve HashedRelation for key fit within Long

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

Currently, we use java HashMap for HashedRelation if the key could fit within a Long. The java HashMap and CompactBuffer are not memory efficient, the memory used by them is also accounted accurately.

This PR introduce a LongToUnsafeRowMap (similar to BytesToBytesMap) for better memory efficiency and performance.
## How was this patch tested?

Updated existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -321,144 +302,422 @@ private[joins] object UnsafeHashedRelation {
...
+}
+
 /**
- * An interface for a hashed relation that the key is a Long.
+ * An hash map mapping from key of Long to UnsafeRow.
  */
-private[joins] trait LongHashedRelation extends HashedRelation {
-  override def get(key: InternalRow): Iterator[InternalRow] = {
-    get(key.getLong(0))
+final class LongToUnsafeRowMap(var mm: TaskMemoryManager, capacity: Int)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: can you describe the physical data layout in the comments here?

**Comment by yhuai** [comment]
wrong jira?

**Comment by sameeragarwal** [comment]
These changes LGTM. It'd be great to add comments that (a) describe the physical layout of the entire data structure (b) explain why lookups are faster (cheaper hash/compare etc.)

**Comment by davies** [comment]
Merging this into master, thanks!

**Pull request closed without merging**